drivers/flash/Kconfig

```diff
@@ -23,6 +23,18 @@ config FLASH_HAS_PAGE_LAYOUT
 	  This option is enabled when the SoC flash driver supports
 	  retrieving the layout of flash memory pages.
 
+config FLASH_ANY_HAS_RESTRICTED_WRITE
+	bool
+	help
+	  This option is enabled by any flash driver with write restrictions,
+	  i.e. requires erase before being written.
+
+config FLASH_ANY_HAS_UNRESTRICTED_WRITE
+	bool
+	help
+	  This option is enabled by any flash driver without write restrictions,
+	  i.e. write does not require an area being erased.
```

Comment:
This is a hardware feature, why is it being described in Kconfig? And can you explain how this would work with SFDP?

Reply:
> This is a hardware feature, why is it being described in Kconfig? 

At the moment this feature is considered to be part of the `compatible` and the property (which is reflecting what kind of devices are available, hence the `_ANY_`) is being set by the driver Kconfig. It could be reflected in dts but this would more-or-less make it configurable. If it is deemed required that this should also be configurable it can be added as a devicetree property. This might require some macrobatics to set the property to the correct value, but should be doable.

> And can you explain how this would work with SFDP?

I suppose you are referring to the fact that it might be needed to discover the property from SFDP. This could be done as:
1. If a compatible needs to support both `restricted_write` and `unrestricted_write` the driver Kconfig enables both properties, the driver bool is then used in subsystems that can take advantage of it. This of course would eliminate any possible code optimizations as it is considered not known at compile time what the properties are. For systems that are not taking advantage of it nothing changes as they are not using the driver bool.
2. To enable back the optimization a property could be added to the devicetree description that forces a certain operation mode. This devicetree property can than be used in the Kconfig to only enable the expected operation mode. To avoid runtime errors a runtime check should be made to make sure that the forced mode is supported by the hardware.

This is only one way, other methods that would "split up" the driver into several device types would also be possible.